src/install_pypi/plan/providers.rs

```diff
@@ -36,18 +37,62 @@ pub trait CachedDistProvider<'a> {
         &mut self,
         name: &'a uv_normalize::PackageName,
         version: uv_pep440::Version,
+        expected_hash: Option<&rattler_lock::PackageHashes>,
     ) -> Option<CachedRegistryDist>;
 }
 
+/// Check if a hash digest matches the expected package hash
+fn hash_matches_expected(
+    hash: &HashDigest,
+    algorithm: HashAlgorithm,
+    expected: &rattler_lock::PackageHashes,
+) -> bool {
+    match (expected, algorithm) {
+        (rattler_lock::PackageHashes::Sha256(expected_sha256), HashAlgorithm::Sha256) => {
+            format!("{:x}", expected_sha256) == hash.to_string()
+        }
+        (rattler_lock::PackageHashes::Md5(expected_md5), HashAlgorithm::Md5) => {
+            format!("{:x}", expected_md5) == hash.to_string()
+        }
+        (rattler_lock::PackageHashes::Md5Sha256(expected_md5, expected_sha256), algo) => match algo
+        {
+            HashAlgorithm::Sha256 => format!("{:x}", expected_sha256) == hash.to_string(),
+            HashAlgorithm::Md5 => format!("{:x}", expected_md5) == hash.to_string(),
+            _ => false,
+        },
+        _ => false,
+    }
+}
+
 impl<'a> CachedDistProvider<'a> for RegistryWheelIndex<'a> {
     fn get_cached_dist(
         &mut self,
         name: &'a uv_normalize::PackageName,
         version: uv_pep440::Version,
+        expected_hash: Option<&rattler_lock::PackageHashes>,
     ) -> Option<CachedRegistryDist> {
-        let index = self
-            .get(name)
-            .find(|entry| entry.dist.filename.version == version);
+        let index = self.get(name).find(|entry| {
+            // Check version matches
+            if entry.dist.filename.version != version {
+                return false;
+            }
+
+            // If we have an expected hash, verify it matches
+            if let Some(expected) = expected_hash {
+                // Check if any of the cached hashes match the expected hash
+                let has_matching_hash = entry
+                    .dist
+                    .hashes
+                    .iter()
+                    .any(|hash| hash_matches_expected(hash, hash.algorithm(), expected));
```

Comment:
Not sure what the convention is, but I would expect *all* hashes to match, not just one!